standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+- extend minimal .Net Core base image
+- set a fixed major version of the base image
+- set the `ASPNETCORE_ENVIRONMENT` environment variable to production in parent
+- run as the `www-data` user and not `root`
+- ensure the working directory is owned by the www-data user - e.g. /app
+- a remote debugger should be added to the SDK file to aid development and testing
+  ```
+  RUN apt-get update \
+  && apt-get install -y --no-install-recommends unzip \
+  && curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg
+  ```
+
+## Container scanning
+Container scanning tools collate vulnerability reports from a number of sources to determine security issues within operating system kernels, libraries, and software packages.
+
+Containers should be regularly scanned using tools such as Clair, Anchore Engine, Aqua Trivy or JFrog XRay.
+
+A comparison of the different tools is available in [Confluence](https://eaflood.atlassian.net/wiki/spaces/FPS/pages/1408172407/Container+Image+Scanning+Tools)
+
+## Multi stage builds
+Dockerfiles should implement multi stage builds to allow different container configuration and behaviour to be created for development, test and production.
+
+Below is an example multi stage build which is intended to use the Future Farming and Countryside (FFC) Node.js parent image.
+
+```
+# Base stage installs production dependencies
+ARG REGISTRY=562955126301.dkr.ecr.eu-west-2.amazonaws.com
+ARG BASE_VERSION=1.0.0
+ARG DEV_VERSION=1.0.0
+FROM $REGISTRY/ffc-node-base:$BASE_VERSION as base
+ARG PORT=3000
+ENV PORT ${PORT}
+USER node
+WORKDIR /home/node
+COPY --chown=node:node package*.json ./
+RUN npm ci
+
+# Development stage installs devDependencies, builds app from source and declares a file watcher as the default command
+FROM $REGISTRY/ffc-node-development:$DEV_VERSION AS development
+EXPOSE ${PORT} 9229 9230
+USER node
+WORKDIR /home/node
+COPY --from=base --chown=node:node /home/node/package*.json ./
+COPY --from=base --chown=node:node /home/node/node_modules ./node_modules
+RUN npm install
+COPY --chown=node:node app/ ./app/
+RUN npm run build
+CMD [ "npm", "run", "start:watch" ]
+
+# Test stage copied in Jest configuration and declares the test task as the default command
+FROM development AS test
+USER node
+WORKDIR /home/node
+COPY --chown=node:node jest.config.js ./jest.config.js
+COPY --chown=node:node test/ ./test/
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production
+USER node
+WORKDIR /home/node
+EXPOSE ${PORT}
+COPY --from=development /home/node/app/ ./app/
+CMD [ "node", "app/index" ]
+```
+
+## Docker Compose
+Docker compose can be used for local development and within a CI/CD pipeline to build docker containers.
+
+### Use override files to reduce duplication
+Additional settings can be applied to a docker compose file by using override files.
```

Comment:
I do agree with this approach to reduce duplication, but I am a little concerned we're creating quite a complicated environment with multi-stage builds in our Dockerfiles and overridden docker-compose files.  The number of permutations points to a high degree of variability in what is actually being built/tested at any given point.

Not sure I've got a good answer beyond "just pick one" but I'd welcome further discussion around this.

Reply:
The two are intended to work in conjunction rather than one being a substitute for another.

ie you'd have a test override file that targets the test build stage.  The compose file just means you don't need  to write a huge `docker` command.  You could do that but I wouldn't suggest it be the standard.

An alternative approach is to have multiple dockerfiles instead of one with multi stage builds, then you'd have a 1-to-1 relationship with docker-compose files.  Though it does result in a lot of repetition and it's easy to miss bits when making changes.